sequencer/src/block.rs

```diff
@@ -81,45 +69,147 @@ pub struct Metadata {
+        self.root
+            .serialize_with_mode(&mut comm_bytes, ark_serialize::Compress::Yes)
+            .unwrap();
+        RawCommitmentBuilder::new(&Self::tag())
+            .var_size_field("root", &comm_bytes)
+            .finalize()
+    }
 
-    #[serde(with = "nmt_serializer")]
-    pub(crate) transaction_nmt: TransactionNMT,
+    fn tag() -> String {
+        "NMTROOT".into()
+    }
 }
 
-impl Block {
-    pub fn header(&self) -> Header {
-        Header {
-            metadata: self.metadata.clone(),
-            transactions_root: NMTRoot {
-                root: self.transaction_nmt.commitment().digest(),
-            },
+impl BlockHeader for Header {
+    type Payload = Payload;
+
+    fn new(payload_commitment: VidCommitment, metadata: NMTRoot, parent_header: &Self) -> Self {
+        Self::from_parent(
+            payload_commitment,
+            metadata,
+            Some(parent_header.height),
+            parent_header.timestamp,
+        )
+    }
+
+    fn genesis() -> (
+        Self,
+        Self::Payload,
+        <Self::Payload as BlockPayload>::Metadata,
+    ) {
+        let (payload, metadata) = Payload::genesis();
+        let payload_commitment = vid_commitment(&payload.encode().unwrap().collect());
+        let header = Self::from_parent(payload_commitment, metadata, None, 0);
+        (header, payload, metadata)
+    }
+
+    fn block_number(&self) -> u64 {
+        self.height
+    }
+
+    fn payload_commitment(&self) -> VidCommitment {
+        self.payload_commitment
+    }
+
+    fn metadata(&self) -> NMTRoot {
+        self.transactions_root
+    }
+}
+
+impl Header {
+    fn from_parent(
+        payload_commitment: VidCommitment,
+        transactions_root: NMTRoot,
+        parent_height: Option<u64>,
+        parent_timestamp: u64,
+    ) -> Self {
+        // The HotShot APIs should be redesigned so that
+        // * they are async
+        // * new blocks being created have access to the application state, which in our case could
+        //   contain an already connected L1 client.
+        // For now, as a workaround, we will create a new L1 client based on environment variables
+        // and use `block_on` to query it.
+        let l1 = if let Some(l1_client) = &*L1_CLIENT {
+            block_on(l1_client.snapshot())
+        } else {
+            // For unit testing, we may disable the L1 client and use mock L1 blocks instead.
+            L1Snapshot {
+                finalized: None,
+                head: 0,
+            }
+        };
+
+        // Increment height.
+        let height = match parent_height {
+            Some(height) => height + 1,
+            None => 0,
+        };
+
+        // Sample a timestamp, ensuring that it does not decrease.
+        let mut timestamp = OffsetDateTime::now_utc().unix_timestamp() as u64;
+        if timestamp < parent_timestamp {
+            tracing::warn!(
+                "Espresso timestamp {timestamp} behind parent {}",
```

Comment:
We are emitting a warning here but not giving instructions what to check or do about it. Would it make sense to ask the operator to check if their servers clock is synced and if yes, ignore the warning?